pkg/controller/machine.go

```diff
@@ -307,43 +303,7 @@ func (c *Controller) syncHandler(key string) error {
 
 	// Delete machine
 	if machine.DeletionTimestamp != nil && sets.NewString(machine.Finalizers...).Has(finalizerDeleteInstance) {
-		if c.deleteProviderInstance(prov, machine); err != nil {
-			if machine, err = c.updateMachineError(machine, machinev1alpha1.DeleteMachineError, err.Error()); err != nil {
-				return fmt.Errorf("failed to update machine error after failed delete: %v", err)
-			}
-			return fmt.Errorf("failed to delete machine at cloudprovider: %v", err)
-		}
-
-		//Check that the instance has really gone
-		if providerInstance, err = c.getProviderInstance(prov, machine); err == nil {
-			if sets.NewString(string(instance.StatusDeleted), string(instance.StatusDeleting)).Has(string(providerInstance.Status())) {
-				glog.V(4).Infof("deletion of instance %s got triggered. Waiting until it fully disappears", providerInstance.ID())
-				c.workqueue.AddAfter(machine.Name, deletionRetryWaitPeriod)
-				return nil
-			}
-			glog.V(2).Infof("instance %s of machine %s is not deleted yet", providerInstance.ID(), machine.Name)
-			c.workqueue.AddAfter(machine.Name, deletionRetryWaitPeriod)
-			return nil
-		} else if err != cloudprovidererrors.ErrInstanceNotFound {
-			return fmt.Errorf("failed to get instance for machine %s after the delete got triggered", machine.Name)
-		}
-
-		glog.V(4).Infof("Deleted machine %s at cloud provider", machine.Name)
-		// Delete delete instance finalizer
-		finalizers := sets.NewString(machine.Finalizers...)
-		finalizers.Delete(finalizerDeleteInstance)
-		machine.Finalizers = finalizers.List()
-		if machine, err = c.updateMachine(machine); err != nil {
-			return fmt.Errorf("failed to update machine after removing the delete instance finalizer: %v", err)
-		}
-		glog.V(4).Infof("Removed delete finalizer from machine %s", machine.Name)
-
-		// Remove error message in case it was set
-		if machine, err = c.clearMachineErrorIfSet(machine, machinev1alpha1.DeleteMachineError); err != nil {
-			return fmt.Errorf("failed to update machine after removing the delete error: %v", err)
-		}
-
-		return nil
+		return c.deleteMachineAndProviderInstance(prov, providerInstance, machine)
```

Comment:
I decided to move deletion logic to a separate function to reduce cyclomatic complexity